src/utils/units.hpp

```diff
@@ -0,0 +1,142 @@
+
+  // Copy constructor must be marked with KOKKOS_FUNCTION
+  KOKKOS_FUNCTION
+  Units(const Units &other) {}
+
+  // Return unit system
+  KOKKOS_INLINE_FUNCTION
+  UnitSystem GetUnitSystem() const { return unit_system_; }
+
+  // Unit conversions
+  KOKKOS_INLINE_FUNCTION
+  Real GetLengthCodeToPhysical() const { return length_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetLengthPhysicalToCode() const { return 1. / length_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetTimeCodeToPhysical() const { return time_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetTimePhysicalToCode() const { return 1. / time_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetMassCodeToPhysical() const { return mass_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetMassPhysicalToCode() const { return 1. / mass_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetEnergyCodeToPhysical() const { return energy_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetEnergyPhysicalToCode() const { return 1. / energy_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetNumberDensityCodeToPhysical() const { return number_density_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetNumberDensityPhysicalToCode() const { return 1. / number_density_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetMassDensityCodeToPhysical() const { return mass_density_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetMassDensityPhysicalToCode() const { return 1. / mass_density_; }
+
+ private:
+  // Unit conversion factors from code to physical units
+  // e.g. length_ has units of cm when using CGS as physical unit system
+  Real length_;
+  Real time_;
+  Real mass_;
+
+  Real energy_;
+  Real number_density_;
+  Real mass_density_;
+  Real temperature_;
+
+  UnitSystem unit_system_;
+};
+
+class Constants {
+ public:
+  Constants() = delete;
+
+  KOKKOS_FUNCTION
+  Constants(Units &units);
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetGPhysical() const { return G_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetGCode() const { return G_code_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetKBPhysical() const { return kb_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetKBCode() const { return kb_code_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetCPhysical() const { return c_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetCCode() const { return c_code_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetHPhysical() const { return h_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetHCode() const { return h_code_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetMsolarPhysical() const { return Msolar_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetMsolarCode() const { return Msolar_code_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetAUPhysical() const { return AU_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetAUCode() const { return AU_code_; }
+
+ private:
+  // Physical constants in physical units
+  Real G_;      // Gravitational constant
+  Real kb_;     // Boltzmann constant
+  Real c_;      // Speed of light
+  Real h_;      // Planck constant
+  Real Msolar_; // Solar mass
+  Real AU_;     // Astronomical unit
```

Comment:
Other constants I'd like to see are:

- [ ]  `amu` and/or `mp` (cv = kb/(mu * amu *(g-1)))
- [ ]   `ev`
- [ ]  `year` 

At some point we'll also want  electron mass, but not at the moment. 

There's also `pc` if we want to be supportive AGN disks

Reply:
Added `amu`, `ev`, `year`, `pc`, good suggestions :+1: